### PR TITLE
[inductor] Speed up batches of Muon updates

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #194181
* #194180
* #194179
* #194178
* #194177

Human commentary from the author: "and add architecture descriptions and examples to each PR to make it clear"

> AI-assisted change description:
>
> ## Summary
>
> Recognize Muon's complete Newton-Schulz program inside `foreach_map`, group
> compatible matrices, and execute profitable groups through the reusable
> symmetric plans. Unsupported or unprofitable lanes retain the ordinary
> reference calculation.
>
> ## Architecture
>
> The lowering has three stages:
>
> ```text
> foreach_map body graph
>          |
>          v
> exact Muon matcher
> - BF16 conversion and optional transpose
> - Frobenius normalization and epsilon clamp
> - repeated Gram / polynomial / update sequence
> - identical coefficients and step count across lanes
>          |
>          v
> _MuonForeachPlan
> - bucket by device and compatible dimensions
> - choose sequential, packed, grouped, or reference execution
> - cap large chunks to control workspace and launch shape
>          |
>          v
> inductor::grouped_muon custom op
> - cache by tensor metadata, CUDA stream, coefficients, and steps
> - serialize use of shared plan workspaces
> - return normal contiguous BF16 tensors
> ```
>
> The matcher accepts either an ordinary `X @ X.T` Gram node or the symmetric
> custom op introduced earlier in the stack. It rejects partial matches,
> differing coefficients, inconsistent transpose behavior, or intermediates
> that do not form the complete Newton-Schulz chain.
>
> Profitability checks require CuTeDSL, SM100/SM110, BF16 inputs, and measured
> shape/count thresholds. Small groups and unsupported cases execute the
> reference tensor program. The process-wide LRU holds at most 32 plans.
> Cache keys include current CUDA stream IDs, and each plan has a run lock
> because its output and workspace tensors are intentionally reused.
>
> ## Example
>
> ```python
> params = [
>     torch.randn(1536, 5120, device="cuda", requires_grad=True)
>     for _ in range(4)
> ]
> optimizer = torch.optim.Muon(params, lr=1e-3, ns_steps=5)
> for param in params:
>     param.grad = torch.randn_like(param)
>
> compiled_step = torch.compile(optimizer.step, fullgraph=True)
> compiled_step()  # creates and caches a grouped plan
> compiled_step()  # reuses compiled code, outputs, metadata, and workspaces
> ```
>
> The user-facing optimizer API remains unchanged. The custom operation is an
> internal lowering target selected only after the full Muon pattern and
> support checks succeed.
>
> ## Benchmark results
>
> Median steady-state BF16 optimizer-step times on an NVIDIA GB200, including
> momentum, normalization, five Newton-Schulz iterations, and parameter
> updates:
>
> | representative rank workload | functional baseline | compiled grouped path | speedup |
> |---|---:|---:|---:|
> | Qwen TP8/PP8, 56 weights | 40.424 ms | 29.340 ms | 1.378x |
> | GLM PP16/EP40, 120 weights | 119.049 ms | 89.066 ms | 1.337x |
>
> The Qwen case saved 11.083 ms per optimizer step and the GLM case saved
> 29.982 ms. Compilation time is excluded; these are development benchmarks,
> not performance guarantees.
>
> ## Test Plan
>
> ```text
> CUDA_VISIBLE_DEVICES=0 TORCHINDUCTOR_CACHE_DIR=/tmp/codex-pytorch2-final-symmetric-cache .venv/bin/python test/inductor/test_symmetric_mm.py -v
> CUDA_VISIBLE_DEVICES=1 .venv/bin/python test/test_optim.py -k Muon -v
> source .venv/bin/activate && lintrunner -a
> ```
>
> Authored with an AI assistant.

### CODE REVIEW

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (architecture, correctness, simplification, generalization, and GPT-5.6 final approval). This is AI-generated — please verify findings manually.*

---

## Verified Issues
- **[high] Matcher drops side effects** — Location: `torch/_inductor/kernel/muon.py:91-195`. Trigger: a matching `foreach_map` body also contains a failing `torch._assert_async`. Failure sequence: the matcher validates only the returned expression, then replaces the entire subgraph, omitting the impure assertion node. Impact: compiled execution silently skips the required assertion. Reviewers: Correctness.
- **[high] Python-only custom op produces unloadable AOTI packages** — Location: `torch/_inductor/lowering.py:1057-1065`, `torch/_inductor/kernel/muon.py:333-373`. Trigger: AOTI-package a matching `foreach_map` and load it in a fresh process. Failure sequence: lowering emits `inductor::grouped_muon`, but its schema and implementation exist only after importing the Python module; the AOTI proxy executor cannot find that schema. Impact: package loading fails. Reviewers: Correctness, Generalization.
- **[medium] Cache eviction happens after workspace allocation** — Location: `torch/_inductor/kernel/muon.py:355-361`. Trigger: 32 large plans fill the cache and a 33rd plan is requested when memory would suffice only after evicting the oldest plan. Failure sequence: `_MuonForeachPlan` allocates all new GPU workspaces before `popitem()` runs. Impact: an avoidable CUDA OOM; completed plans can also retain tens of GiB after their originating optimizer or graph is gone. Reviewers: Architecture, Correctness, Generalization.
- **[medium] Replacement changes output layout** — Location: `torch/_inductor/kernel/muon.py:324,368-373`. Trigger: return a matched result for a tall matrix. Failure sequence: eager execution returns the final transpose view, which is noncontiguous, while the replacement clones it into contiguous storage and its fake implementation advertises a contiguous tensor. Impact: eager and compiled results disagree on strides and `is_contiguous()`, changing subsequent layout-sensitive behavior. Reviewers: Generalization.
- **[low] CuTe-dependent tests are not dependency-gated** — Location: `test/inductor/test_symmetric_mm.py:308-566`. Trigger: run on SM100/SM110 without CuTeDSL installed. Failure sequence: tests pass the architecture check, then either import `grouped_symmetric_mm` and fail or receive no optimized plan while asserting that one exists. Impact: deterministic test failures instead of skips on a supported optional-dependency configuration. Reviewers: Generalization.
- **[low] Muon test helper omits cross-stream synchronization** — Location: `test/inductor/test_symmetric_mm.py:35-47`. Trigger: default-stream tensor initialization is still pending when the newly created nonblocking stream starts. Failure sequence: the helper enters that stream without `wait_stream()` and reads or updates those tensors. Impact: nondeterministic incorrect test results or failures. Reviewers: Correctness.
